File: schemas/report.py
"""智能报告模块 V2 - API Schema。

这个文件定义的是 FastAPI 请求体/响应体，不是数据库模型。

面试时可以这样讲：
“ORM 负责表结构，Schema 负责接口契约。报告模块 V2 把任务外壳和不同报告类型
的内容结构拆开，前端可以根据 report_type + schema_version 选择渲染器。”
"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import AliasChoices, BaseModel, Field

from services.reporting.registry import REPORT_REGISTRY
from services.reporting.schemas import REPORT_SCHEMA_VERSION


REPORT_TYPES: list[str] = list(REPORT_REGISTRY.keys())


class ReportGenerateRequest(BaseModel):
    """POST /api/v1/reports/generate 请求体。"""

    report_type: str = Field(..., description="报告类型编码")
    report_title: str = Field(
        ...,
        validation_alias=AliasChoices("report_title", "title"),
        min_length=1,
        max_length=255,
        description="报告标题；兼容前端传 title，也兼容后端内部字段 report_title",
    )
    period_start: Optional[date] = Field(default=None, description="统计周期开始")
    period_end: Optional[date] = Field(default=None, description="统计周期结束")
    filters: dict[str, Any] = Field(default_factory=dict, description="报告筛选条件")

    # 这里开启字段名回填，是为了让调用方既可以传前端友好的 title，
    # 也可以传数据库/响应一致的 report_title；服务层仍统一读取 report_title。
    model_config = {"populate_by_name": True}


class ReportTaskResponse(BaseModel):
    """报告任务创建后的轻量响应。"""

    id: int
    report_type: str
    report_title: str
    status: str
    schema_version: int = REPORT_SCHEMA_VERSION
    period_start: Optional[date] = None
    period_end: Optional[date] = None
    trigger_source: str = "manual"
    generated_by: Optional[int] = None
    create_time: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ReportDetailResponse(ReportTaskResponse):
    """报告详情响应。"""

    report_content: Optional[dict[str, Any]] = None
    report_html: Optional[str] = None
    data_quality: Optional[dict[str, Any]] = None
    request_filters: Optional[dict[str, Any]] = None
    aggregated_data_snapshot: Optional[dict[str, Any]] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    started_time: Optional[datetime] = None
    completed_time: Optional[datetime] = None
    retry_of_report_id: Optional[int] = None
    retry_count: int = 0


class ReportTypeResponse(BaseModel):
    """GET /api/v1/reports/types 返回的报告类型契约。"""

    report_type: str
    label: str
    schema_version: int
    allowed_roles: list[str]
    template_name: str
    default_period_rule: str
    available_filters: list[str]
    json_schema: dict[str, Any]


class ReportListResponse(BaseModel):
    items: list[ReportTaskResponse]
    total: int
    page: int
    page_size: int


class ReportScheduleCreate(BaseModel):
    report_type: str
    cron_expression: str = Field(..., description="五段 cron 表达式，例如 0 9 * * 1")
    enabled: int = 1
    timezone: str = "Asia/Shanghai"
    period_rule: str = "previous_week"
    title_template: Optional[str] = None
    filters: dict[str, Any] = Field(default_factory=dict)
    recipients: dict[str, Any] = Field(default_factory=dict)


class ReportScheduleUpdate(BaseModel):
    cron_expression: Optional[str] = None
    enabled: Optional[int] = None
    timezone: Optional[str] = None
    period_rule: Optional[str] = None
    title_template: Optional[str] = None
    filters: Optional[dict[str, Any]] = None
    recipients: Optional[dict[str, Any]] = None


class ReportScheduleResponse(ReportScheduleCreate):
    id: int
    filters: Optional[dict[str, Any]] = None
    recipients: Optional[dict[str, Any]] = None
    created_by: Optional[int] = None
    last_run_time: Optional[datetime] = None
    last_status: Optional[str] = None
    last_error: Optional[str] = None
    next_run_time: Optional[datetime] = None
    create_time: Optional[datetime] = None
    update_time: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ReportActionCreate(BaseModel):
    suggestion_text: str = Field(
        ...,
        validation_alias=AliasChoices("suggestion_text", "suggestion"),
        description="从报告建议转成的行动内容；兼容前端传 suggestion",
    )
    risk_code: Optional[str] = None
    owner_id: Optional[int] = None
    due_time: Optional[datetime] = None
    target_value: Optional[Decimal] = None

    # 与报告标题类似，行动项内部字段叫 suggestion_text，
    # 但前端表单更常用 suggestion；这里统一在 Schema 层做兼容。
    model_config = {"populate_by_name": True}


class ReportActionUpdate(BaseModel):
    suggestion_text: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("suggestion_text", "suggestion"),
    )
    risk_code: Optional[str] = None
    owner_id: Optional[int] = None
    due_time: Optional[datetime] = None
    status: Optional[str] = None
    target_value: Optional[Decimal] = None
    actual_value: Optional[Decimal] = None
    completed_time: Optional[datetime] = None

    model_config = {"populate_by_name": True}


class ReportActionResponse(ReportActionCreate):
    id: int
    report_id: int
    status: str
    actual_value: Optional[Decimal] = None
    completed_time: Optional[datetime] = None
    create_time: Optional[datetime] = None
    update_time: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ApplicationMaterialCreate(BaseModel):
    application_id: int
    student_id: Optional[int] = None
    owner_id: Optional[int] = None
    material_name: str
    required: int = Field(
        default=1,
        validation_alias=AliasChoices("required", "is_required"),
        description="是否必填；兼容前端传 is_required",
    )
    deadline: Optional[date] = None
    submitted_time: Optional[datetime] = None
    status: str = "pending"

    model_config = {"populate_by_name": True}


class ChannelCostCreate(BaseModel):
    channel: str
    cost_date: date = Field(
        ...,
        validation_alias=AliasChoices("cost_date", "spend_date"),
        description="投放/成本日期；兼容前端传 spend_date",
    )
    campaign: Optional[str] = None
    cost_amount: Decimal = Field(
        ...,
        validation_alias=AliasChoices("cost_amount", "cost"),
        description="渠道成本；兼容前端传 cost",
    )

    model_config = {"populate_by_name": True}


class CustomerContractCreate(BaseModel):
    customer_id: int
    lead_id: Optional[int] = None
    channel: Optional[str] = None
    contract_amount: Decimal
    signed_time: Optional[datetime] = None
    status: str = "signed"


class CustomerPaymentCreate(BaseModel):
    contract_id: int
    payment_amount: Decimal
    payment_time: Optional[datetime] = Field(
        default=None,
        validation_alias=AliasChoices("payment_time", "paid_time"),
        description="回款时间；兼容前端传 paid_time",
    )
    status: str = "paid"

    model_config = {"populate_by_name": True}


class CRMLeadStatusUpdate(BaseModel):
    old_status: Optional[str] = None
    new_status: str
    change_reason: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("change_reason", "reason"),
        description="阶段变更原因；兼容前端传 reason",
    )

    model_config = {"populate_by_name": True}


class FeedbackTicketUpdate(BaseModel):
    status: Optional[str] = None
    priority: Optional[str] = None
    category: Optional[str] = None
    owner_id: Optional[int] = None
    first_response_time: Optional[datetime] = None
    resolved_time: Optional[datetime] = None
    satisfaction_score: Optional[int] = None

File: frontend/src/types/report.ts
/**
 * 智能报告相关 TypeScript 类型。
 *
 * 对齐后端：
 *   - schemas/report.py（API 请求/响应 Schema）
 *   - services/reporting/schemas.py（10 类报告内容模型）
 *   - services/reporting/registry.py（报告类型注册表）
 *
 * 所有字段结构已在阶段 0 契约审查中通过 Python 代码确认。
 */

import type { DataQuality, ReportStatus, TriggerSource, PaginatedResponse } from './common'

// ============================================================
//  报告类型元数据
// ============================================================

/** GET /api/v1/reports/types 返回的报告类型定义 */
export interface ReportTypeDefinition {
  report_type: string
  label: string
  schema_version: number
  allowed_roles: string[]
  template_name: string
  default_period_rule: string
  available_filters: string[]
  /** 该报告类型的内容 JSON Schema */
  json_schema: Record<string, unknown>
}

// ============================================================
//  报告生成
// ============================================================

/** POST /api/v1/reports/generate 请求体 */
export interface ReportGenerateRequest {
  report_type: string
  /** 兼容前端传 "title"，后端自动映射为 "report_title" */
  report_title: string
  period_start?: string | null
  period_end?: string | null
  filters?: Record<string, unknown>
}

/** 报告任务创建后的轻量响应（202 Accepted） */
export interface ReportTaskResponse {
  id: number
  report_type: string
  report_title: string
  status: ReportStatus
  schema_version: number
  period_start?: string | null
  period_end?: string | null
  trigger_source: TriggerSource
  generated_by?: number | null
  create_time?: string | null
}

/** 报告列表响应 */
export type ReportListResponse = PaginatedResponse<ReportTaskResponse>

/** GET /api/v1/reports/{id} 详情响应 */
export interface ReportDetailResponse extends ReportTaskResponse {
  report_content?: Record<string, unknown> | null
  report_html?: string | null
  data_quality?: DataQuality | null
  request_filters?: Record<string, unknown> | null
  aggregated_data_snapshot?: Record<string, unknown> | null
  error_code?: string | null
  error_message?: string | null
  started_time?: string | null
  completed_time?: string | null
  retry_of_report_id?: number | null
  retry_count: number
}

// ============================================================
//  报告行动项
// ============================================================

export interface ReportActionCreate {
  /** 兼容前端传 "suggestion"，后端自动映射为 "suggestion_text" */
  suggestion_text: string
  risk_code?: string | null
  owner_id?: number | null
  due_time?: string | null
  target_value?: number | null
}

export interface ReportActionResponse extends ReportActionCreate {
  id: number
  report_id: number
  status: string
  actual_value?: number | null
  completed_time?: string | null
  create_time?: string | null
  update_time?: string | null
}

export interface ReportActionUpdate {
  suggestion_text?: string | null
  risk_code?: string | null
  owner_id?: number | null
  due_time?: string | null
  status?: string | null
  target_value?: number | null
  actual_value?: number | null
  completed_time?: string | null
}

// ============================================================
//  报告定时计划
// ============================================================

export interface ReportScheduleCreate {
  report_type: string
  cron_expression: string
  enabled: number
  timezone: string
  period_rule: string
  title_template?: string | null
  filters: Record<string, unknown>
  recipients: Record<string, unknown>
}

export interface ReportScheduleResponse extends ReportScheduleCreate {
  id: number
  created_by?: number | null
  last_run_time?: string | null
  last_status?: string | null
  last_error?: string | null
  next_run_time?: string | null
  create_time?: string | null
  update_time?: string | null
}

// ============================================================
//  10 类报告内容模型（对齐 services/reporting/schemas.py）
// ============================================================

/** 所有报告内容共享的基础字段 */
export interface BaseReportContent {
  summary: string
  explanation: string
  metric_traces: MetricTrace[]
}

/** 指标追溯信息（对齐后端 MetricTrace） */
export interface MetricTrace {
  metric_name: string
  source_tables: string[]
  formula?: string | null
  filters: Record<string, unknown>
}

/** 申请风险指标 */
export interface ApplicationRiskMetrics {
  total_applications: number
  high_risk_count: number
  medium_risk_count: number
  low_risk_count: number
  overdue_count: number
  missing_material_count: number
}

/** 申请风险明细 */
export interface ApplicationRiskItem {
  application_id: number
  student_id?: number | null
  owner_id?: number | null
  stage: string
  risk_score: number
  risk_level: 'high' | 'medium' | 'low'
  risk_reasons: string[]
  missing_materials: string[]
  next_action?: string | null
}

/** 报告行动建议 */
export interface ReportActionSuggestion {
  owner_id?: number | null
  action: string
  due_date?: string | null
  priority: 'urgent' | 'high' | 'medium' | 'low'
}

export type CustomerOpsContent = BaseReportContent & {
  metrics: Record<string, unknown>
  stage_distribution: Record<string, unknown>[]
  stale_leads: Record<string, unknown>[]
  churn_analysis: Record<string, unknown>[]
}

export type DailySummaryContent = BaseReportContent & {
  metrics: Record<string, unknown>
  key_progress: string[]
  common_risks: string[]
  next_plans: string[]
}

export type WeeklySummaryContent = BaseReportContent & {
  business_sections: Record<string, unknown>
  cross_module_risks: string[]
  management_actions: string[]
}

export type PsychWeeklyContent = BaseReportContent & {
  metrics: Record<string, unknown>
  emotion_trend: Record<string, unknown>[]
  alert_status: Record<string, unknown>[]
  processing_timeliness: Record<string, unknown>
}

export type ComplaintWeeklyContent = BaseReportContent & {
  metrics: Record<string, unknown>
  sla_summary: Record<string, unknown>
  high_frequency_issues: Record<string, unknown>[]
}

export type ApplicationRiskContent = BaseReportContent & {
  metrics: ApplicationRiskMetrics
  risk_items: ApplicationRiskItem[]
  action_checklist: ReportActionSuggestion[]
}

export type SalesFunnelContent = BaseReportContent & {
  funnel_counts: Record<string, number>
  conversion_rates: Record<string, number | null>
  avg_stage_stay_days: Record<string, number | null>
  stalled_leads: Record<string, unknown>[]
  consultant_performance: Record<string, unknown>[]
}

export type ChannelROIContent = BaseReportContent & {
  channel_metrics: Record<string, unknown>[]
  data_quality_warnings: string[]
}

export type ServiceSLAContent = BaseReportContent & {
  sla_overview: Record<string, unknown>
  complaint_sla: Record<string, unknown>[]
  admin_service_sla: Record<string, unknown>[]
  psych_alert_sla: Record<string, unknown>[]
  backlog_aging: Record<string, unknown>[]
}

export type ActionClosureContent = BaseReportContent & {
  metrics: Record<string, unknown>
  overdue_actions: Record<string, unknown>[]
  repeated_issues: Record<string, unknown>[]
  target_achievement: Record<string, unknown>[]
}

/** 报告内容联合类型 */
export type ReportContentUnion =
  | CustomerOpsContent
  | DailySummaryContent
  | WeeklySummaryContent
  | PsychWeeklyContent
  | ComplaintWeeklyContent
  | ApplicationRiskContent
  | SalesFunnelContent
  | ChannelROIContent
  | ServiceSLAContent
  | ActionClosureContent
  | Record<string, unknown> // V1 兼容

// ============================================================
//  报告类型展示配置（本地保留，不可作为数据源）
// ============================================================

/** 报告类型的展示配置（图标、重点色） */
export interface ReportPresentation {
  icon: string
  accent: 'default' | 'warning' | 'info' | 'success' | 'danger'
  description: string
}

/** 10 类报告的展示配置 */
export const REPORT_PRESENTATION: Record<string, ReportPresentation> = {
  customer_ops: { icon: 'Users', accent: 'info', description: '客户阶段分布、停滞线索、流失风险分析' },
  daily_summary: { icon: 'ClipboardList', accent: 'default', description: '员工日报汇总、关键进展、共性问题' },
  weekly_summary: { icon: 'BarChart3', accent: 'info', description: '跨模块综合经营周报、管理行动建议' },
  psych_weekly: { icon: 'Heart', accent: 'danger', description: '心理预警分布、情绪趋势、首次跟进时效' },
  complaint_weekly: { icon: 'MessageSquareWarning', accent: 'warning', description: '投诉分类统计、SLA 时效、高频问题' },
  application_risk: { icon: 'ShieldAlert', accent: 'warning', description: '申请材料风险评分、缺失追踪、行动建议' },
  sales_funnel: { icon: 'Funnel', accent: 'info', description: '销售漏斗转化率、阶段停留、停滞线索' },
  channel_roi: { icon: 'DollarSign', accent: 'success', description: '渠道投放成本、CPL/CAC/ROI 对比' },
  service_sla: { icon: 'Clock', accent: 'warning', description: '服务响应/解决时效、积压老化分析' },
  action_closure: { icon: 'CheckCircle', accent: 'success', description: '行动完成率、逾期追踪、重复问题' },
}
